src/cargo/core/compiler/job_queue.rs

```diff
@@ -764,6 +797,71 @@ impl<'cfg> DrainState<'cfg> {
                 // It doesn't really matter if this fails.
                 drop(cx.bcx.config.shell().status("Finished", message));
             }
+
+            if !self.future_incompat_reports.is_empty() {
+                self.future_incompat_reports.sort_by_key(|r| r.krate);
+
+                let crates_and_versions = self
+                    .future_incompat_reports
+                    .iter()
+                    .map(|r| format!("`{}` v{}", r.krate, r.version))
+                    .collect::<Vec<_>>()
+                    .join(", ");
+
+                drop(cx.bcx.config.shell().warn(&format!("the following crates contain code that will be rejected by a future version of Rust: {}",
+                                                         crates_and_versions)));
+
+                let mut full_report = String::new();
+                // Generate a random UUID
+                let uuid = Uuid::new_v4();
+
+                for report in self.future_incompat_reports {
+                    // Writing to a String can never error
+                    writeln!(&mut full_report, "The crate `{}` v{} currently triggers the following future incompatibility lints:", report.krate, report.version).unwrap();
+                    for item in report.report {
+                        for line in item.diagnostic.rendered.lines() {
+                            // Writing to a String can never error
+                            writeln!(&mut full_report, "> {}", line).unwrap();
+                        }
+                    }
+                }
+
+                match cx.bcx.ws.target_dir().open_rw(
+                    FUTURE_INCOMPAT_FILE,
+                    cx.bcx.config,
+                    "Future incompatibility report",
+                ) {
+                    Ok(report_file) => {
+                        let on_disk_report = OnDiskReport {
+                            uuid,
+                            report: Cow::Borrowed(&full_report),
+                        };
+                        if let Err(e) = serde_json::to_writer(report_file, &on_disk_report) {
+                            drop(cx.bcx.config.shell().warn(&format!(
+                                "Failed to write on-disk future incompat report: {:?}",
+                                e
+                            )));
+                        }
+                    }
+                    Err(e) => {
+                        drop(cx.bcx.config.shell().warn(&format!(
+                            "Failed to open on-disk future incompat report: {:?}",
+                            e
+                        )));
+                    }
+                }
+
+                if self.show_future_incompat_report {
+                    drop_eprint!(cx.bcx.config, "{}", full_report);
+                    drop(cx.bcx.config.shell().note(
+                        &format!("this report can be shown with `cargo describe-future-incompatibilities -Z future-incompat-report --id {}`", uuid)
```

Comment:
Should I keep the `-Z` flag for gating `--future-incompat-report`, or get rid of it entirely?

Reply:
For a brand new subcommand I think it's fine to just say that the subcommand works on nightly and otherwise unconditionally prints an error on stable/beta. I'm not sure that we have a ton of precedent for new unstable subcommands.

Reply:
I meant for the flag pass to `cargo build`. If we don't have a `-Z` flag, then we'll effectively make this on-by-default on Nightly, since we'll start printing the short notice when running `cargo build`.

Reply:
Yes if `cargo build` has a `--future-incompat-report` flag then it needs a `-Z` flag to gate that flag.